### DISCUSSION TITLE
How to replace the DataPortalProxy at an arbitrary time in the app lifecycle

### DISCUSSION THREAD

**Discussion by GreatBarrier86** [MarimerLLC/csla#3478] [Questions]
I have a question on this. In CSLA 5, we had  the `Csla.ApplicationContext.DataPortalProxy` and `Csla.ApplicationContext.DataPortalUrlString` static properties that allowed changing the values at runtime (eg: switching from local to remote at an arbitrary time). 

In principle, it doesn't appear that Dependency Injection supports that concept. Would you say changing the proxy/url is simply not a thing we could do anymore?

**Comment by GreatBarrier86** [comment]
I'm extraordinarily new to using Dependency Injection so i wasn't coming at this from the right angle. It occurs to me that while you can't replace services after you're registered them, there's nothing stopping me from using the Func<> parameter of each of the `AddTransient` calls to return what I need at DP request time to switch back and forth.

Would this be reasonable to you, @rockfordlhotka?

`

                services.AddTransient<HttpProxyOptions>(p =>
                        {
                            return new HttpProxyOptions { DataPortalUrl = CslaUtilities.DataPortalHelper.WebServiceUrl };
                        })
                        .AddTransient<HttpClient>()
                        .AddCsla()
                        .AddTransient<IDataPortalProxy>(p =>
                        {
                            var appCtx = p.GetRequiredService<Csla.ApplicationContext>();
                            if (CslaUtilities.DataPortalHelper.UseRemoteDataPortal)
                            {
                                var httpClient = p.GetRequiredService<HttpClient>();
                                var proxOps = p.GetRequiredService<HttpProxyOptions>();
                                return new CustomHttpProxy(appCtx, httpClient, proxOps);
                            }
                            else
                            {
                                return new LocalProxy(appCtx, new LocalProxyOptions());
                            }
                        });
`

**Reply by rockfordlhotka** [reply]
You are on the right track, but the specific solution won't work - at least not if you use the existing extension method: https://github.com/MarimerLLC/csla/blob/ffa5f2243a1ef72d53379fb59863fa6cf070ad59/Source/Csla/DataPortalClient/HttpProxyExtensions.cs#L25

The reason is that, in the default extension method, the `HttpProxyOptions` are directly created, they aren't created as a service.

I think what you want is to change how the actual `HttpProxy` is created and initialized. Basically DO NOT use the existing `UseHttpProxy` extension method, and instead create your own implementation that sets up the `HttpProxyOptions` as necessary for each call.

**Reply by GreatBarrier86** [reply]
Where am I using UseHttpProxy? Maybe I misunderstand.

**Reply by rockfordlhotka** [reply]
In `Program.cs` on the client, where you call something like `services.AddCsla()` you'll call other configuration elements for CSLA, usually including setting the data portal proxy.

From `BlazorExample`:

```c#
builder.Services.AddCsla(o => o
  .AddBlazorWebAssembly()
  .DataPortal(dpo => dpo
    .EnableSecurityPrincipalFlowFromClient()
    .UseHttpProxy(options => options.DataPortalUrl = "/api/DataPortal")));
```
